### PR TITLE
Logger Destination Refactor

### PR DESCRIPTION
**Pull request by eracah**
Closed. See https://github.com/mosaicml/composer/pull/1416

- [x] add new methods (`log_metrics`, `log_hyperparameters`) to `LoggerDestination`
- [x] add corresponding methods to `Logger`
- [x] Add corresponding methods to WandB logger
- [x] Add corresponding methods to TB logger
- [x] Add corresponding methods to File logger
- [x] Add corresponding methods to Progress Bar logger
- [x] Add corresponding methods to in memory logger
- [x] refactor all `Logger` calls in `Trainer` to use new methods
- [x] refactor all `Logger` calls in `Algorithms` to use new methods
- [x] refactor all `Logger` calls elsewhere to use new methods
- [x] remove `Logger.data()` method
- [x] remove all `data_fit` 
- [x] remove all `data_batch`
- [x] remove all `data_epoch`
- [x] Remove `log_data` methods everywhere
- [x] Test progress bar in notebook
- [ ] Test progress bar on cluster/k8's
- [x] Remove loglevel everywhere
- [ ] update docs with changes
- [ ] update notebooks
- [ ] update tests
- [x] add log_traces
- [ ] test with all loggers on on the cluster

### CODE REVIEW

**Comment by eracah** [comment]
It's a little rough around the edges, but I figured I'd get a little feedback first.

File: composer/loggers/logger.py [hunk 1]
```diff
@@ -79,18 +79,18 @@ def __init__(
         self.destinations = ensure_tuple(destinations)
         self._state = state
 
-    def data(self, log_level: Union[str, int, LogLevel], data: Dict[str, Any]) -> None:
-        """Log data to the :attr:`destinations`.
+    def log_traces(self, traces: Dict[str, Any]):
+        for destination in self.destinations:
+            destination.log_traces(traces)
```
Comment (mvpatel2000) [inline composer/loggers/logger.py:84]: Why does this have to be different from metrics?

File: composer/loggers/wandb_logger.py [hunk 2]
```diff
@@ -98,11 +99,16 @@ def __init__(
...
-    def log_data(self, state: State, log_level: LogLevel, data: Dict[str, Any]):
+    def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
         import wandb
-        del log_level  # unused
-        if self._enabled:
-            wandb.log(data, step=int(state.timestamp.batch))
+        wandb.config.update(hyperparameters)
+
+    def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) -> None:
+        import wandb
```
Comment (mvpatel2000) [inline composer/loggers/wandb_logger.py:107]: Why are these `import wandb` in line?

File: composer/loggers/wandb_logger.py [hunk 3]
```diff
@@ -137,6 +143,7 @@ def init(self, state: State, logger: Logger) -> None:
         if self._enabled:
             wandb.init(**self._init_kwargs)
             atexit.register(self._set_is_in_atexit)
+        print(wandb.run.dir)
```
Comment (mvpatel2000) [inline composer/loggers/wandb_logger.py:146]: flagging for deletion

File: composer/trainer/trainer.py [hunk 4]
```diff
@@ -1537,6 +1528,12 @@ def _train_loop(self) -> None:
                     # `now` is actually in the past, but want to include the time it takes to perform this reduction
                     last_wct = now
 
+                    if self.train_metrics is not None:
+                        self._compute_and_log_metrics(
+                            dataloader_label='train',
+                            metrics=self.train_metrics,
+                        )
+
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1536]: Why is this moved? Is it because you want the older timestamp?

File: tests/callbacks/test_image_visualizer.py [hunk 5]
```diff
@@ -18,7 +18,7 @@
 except ImportError:
     _WANDB_INSTALLED = False
 
-
+#@pytest.mark.skip
```
Comment (mvpatel2000) [inline tests/callbacks/test_image_visualizer.py:21]: marking for deletion

**Review by mvpatel2000 — COMMENTED** [review]
This looks great! Only major comment in initial pass: can you explain why `log_traces` has to be different? 

Also, it'd be awesome if you could include pictures of testing when you get to that point.

File: composer/loggers/logger.py (hunk 1 above)
Comment (eracah) [reply composer/loggers/logger.py:84]: It's for logging [engine traces,](https://github.com/mosaicml/composer/blob/dev/composer/core/engine.py#L310) which we want in the `FileLogger`, but we don't want in `WandBLogger`, so `WandBLogger` doesn't implement `log_traces`

File: composer/loggers/wandb_logger.py (hunk 2 above)
Comment (eracah) [reply composer/loggers/wandb_logger.py:107]: because it doesn't get imported anywhere else. We don't want it as a top-level import because the user may not have it installed

File: composer/trainer/trainer.py (hunk 4 above)
Comment (eracah) [reply composer/trainer/trainer.py:1536]: yes exactly. train_metrics had an off by one error

File: tests/callbacks/test_image_visualizer.py (hunk 5 above)
Comment (eracah) [reply tests/callbacks/test_image_visualizer.py]: good catch! done

File: composer/loggers/wandb_logger.py (hunk 3 above)
Comment (eracah) [reply composer/loggers/wandb_logger.py]: nice spot. done

**Comment by eracah** [comment]
This PR is moved to https://github.com/mosaicml/composer/pull/1416
